packages/server/src/leaderboard/calculate.ts

```diff
@@ -0,0 +1,143 @@
+import { getScore } from '../util/scores'
+import { WorkerRequest, WorkerResponse, ChallengeInfo, UserInfo, GraphEntry } from './types'
+
+export default ({ solves, users, graphUpdate, challenges, config }: WorkerRequest): WorkerResponse => {
+  const challengeInfos = new Map<ChallengeInfo['id'], ChallengeInfo>()
+  for (let i = 0; i < challenges.length; i++) {
+    const challenge = challenges[i]
+    challengeInfos.set(challenge.id, {
+      id: challenge.id,
+      tiebreakEligible: challenge.tiebreakEligible,
+      solves: 0,
+      score: 0
+    })
+  }
+  const userInfos = new Map<UserInfo['id'], UserInfo>()
+  for (let i = 0; i < users.length; i++) {
+    const user = users[i]
+    userInfos.set(user.id, {
+      id: user.id,
+      name: user.name,
+      division: user.division,
+      score: 0,
+      lastSolve: 0,
+      solvedChallengeIds: []
+    })
+  }
+
+  const graphSampleTime = config.leaderboard.graphSampleTime
+  const calcSamples = ({ start, end }: { start: number, end: number }): number[] => {
+    const samples = []
+    const sampleStart = Math.ceil(start / graphSampleTime) * graphSampleTime
+    const sampleEnd = Math.floor(end / graphSampleTime) * graphSampleTime
+
+    for (let sample = sampleStart; sample <= sampleEnd; sample += graphSampleTime) {
+      samples.push(sample)
+    }
+    return samples
+  }
+
+  let lastIdx = 0
+  const calcScores = (sample: number): void => {
+    for (; lastIdx < solves.length; lastIdx++) {
+      const challId = solves[lastIdx].challengeid
+      const userId = solves[lastIdx].userid
+      const createdAt = solves[lastIdx].createdat.valueOf()
+
+      if (createdAt > sample) {
+        break
+      }
+
+      const challengeInfo = challengeInfos.get(challId)
+      if (challengeInfo === undefined) {
+        continue
+      }
+      const userInfo = userInfos.get(userId)
+      if (userInfo === undefined) {
+        continue
+      }
+
+      challengeInfo.solves++
+
+      if (challengeInfo.tiebreakEligible) {
+        userInfo.lastSolve = createdAt
+      }
+
+      userInfo.solvedChallengeIds.push(challId)
+    }
+
+    let maxSolves = 0
+    for (const [, challengeInfo] of challengeInfos) {
+      if (challengeInfo.solves > maxSolves) {
+        maxSolves = challengeInfo.solves
+      }
+    }
+
+    for (let i = 0; i < challenges.length; i++) {
+      const challenge = challenges[i]
+      const challengeInfo = challengeInfos.get(challenge.id) as ChallengeInfo
```

Comment:
This is more clearly expressed as a non-null assertion + an eslint disable comment. Also, why is this not implemented as a loop over `challengeInfos` to avoid nulls entirely?

Reply:
`points.min` and `points.max` are not available on `ChallengeInfo`, only on `Challenge`. Because it's not  possible to efficiently retrieve the `Challenge` for a given `ChallengeInfo`, we iterate over `challenges` instead of `challengeInfos`.